### ISSUE TITLE
Can't find the game when reloading the game from browser

### ISSUE THREAD

**Issue by akh64bit** [giongto35/cloud-game#189]
I am running the game locally on chrome.
When I go to localhost:8000 for the first time after launching the docker, the game launches successfully. Here is the logs from the console. 
[input] keyboard has been initialized
log.js?v=3:9 [input] joystick has been initialized
log.js?v=3:9 [input] touch input has been initialized
log.js?v=3:9 [ws] <- open connection
log.js?v=3:9 [ws] -> setting ping interval to 2000ms
log.js?v=3:9 [ping] <-> {http://localhost:9000/echo: 5}
log.js?v=3:9 [rtcp] <- received stunturn from worker [{"urls":"stun:stun.l.google.com:19302"},{"urls":"stun:127.0.0.1:3478"},{"urls":"turn:127.0.0.1:3478","username":"root","credential":"root"}]
log.js?v=3:9 Added voice track
log.js?v=3:9 [rtcp] ice gathering
log.js?v=3:9 [rtcp] got ice candidate: {"candidate":"candidate:3792981046 1 udp 2122260223 10.0.0.26 53547 typ host generation 0 ufrag 85jv network-id 1 network-cost 10","sdpMid":"0","sdpMLineIndex":0}
log.js?v=3:9 [rtcp] ice gathering completed
log.js?v=3:9 [rtcp] <- iceConnectionState: connected
log.js?v=3:9 [rtcp] connected...
log.js?v=3:9 [rtcp] <- iceConnectionState: disconnected
log.js?v=3:9 [rtcp] disconnected...

However when I reload the game from browser, I can't see any game in the game window. Here are the logs when it fails to reload.
[input] keyboard has been initialized
log.js?v=3:9 [input] joystick has been initialized
log.js?v=3:9 [input] touch input has been initialized
log.js?v=3:9 [ws] <- open connection
log.js?v=3:9 [ws] -> setting ping interval to 2000ms

I have to run the docker again for launching the game again.
Does anyone know of the issue.

Thanks in advance for your help.

**Comment by sergystepanov** [comment]
Hi @akh64bit,

Could you please show your Docker logs with coordinator / worker output during this lunch / relunch?

**Comment by giongto35** [comment]
@akh64bit 
It sometimes happens because the connection cleaning process is not properly implemented (code bug issue). The worker is panic due to "sending to a closed channel". The bug can be reproduced by keep reloading. I didn't have time to fully fix that.
However, it's still better if you can grab it from the worker log. Don't know if there is easy way to get log from docker now.

**Comment by sergystepanov** [comment]
> The worker is panic due to "sending to a closed channel". The bug can be reproduced by keep reloading. I didn't have time to fully fix that.
Is that specific for Docker? I couldn't reproduce it on a native nix or win, or even WSL2.

**Comment by giongto35** [comment]
@sergystepanov It should not happen frequently as I tried to fix it before. Only one case left when I keep opening the game and reload. (Listed below)
Yes, It is rare, so It might not be the same with the issue he is facing. 
```panic: send on closed channel

goroutine 550 [running]:
github.com/giongto35/cloud-game/pkg/webrtc.(*WebRTC).StartClient.func7(0xc0006283f0, 0xc0005ba120)
        /home/thanh/Desktop/Code/cloud-game/pkg/webrtc/webrtc.go:213 +0x120
created by github.com/pion/webrtc/v2.(*PeerConnection).onTrack
        /home/thanh/go/pkg/mod/github.com/pion/webrtc/v2@v2.2.0/peerconnection.go:265 +0x117
exit status 2
```

**Comment by sergystepanov** [comment]
I see. If you'd run workers as a service or use Docker's restart-always option (worker as a separate container) that shouldn't matter to some extent (:
Interesting crash in a sessions map you got there.
I think this ticket is more likely about some core crash on exit as mGBA.

**Comment by akh64bit** [comment]
@sergystepanov here are the docker logs after I have launched the docker for the first time.
`I0602 22:47:44.483814       6 main.go:37] Initializing coordinator server
I0602 22:47:44.483850       6 main.go:38] Coordinator configs {8800 http://localhost:8000   {6601 /coordinator false false}}
I0602 22:47:44.485844       6 coordinator.go:45] Starting monitoring server for coordinator
I0602 22:47:44.486987       6 coordinator.go:103] Initializing Coordinator Server
I0602 22:47:44.488093       6 monitoring.go:81] Monitoring server is disabled via config
I0602 22:47:44.655008       7 main.go:29] Initializing worker server
I0602 22:47:44.656448       7 worker.go:47] Starting monitoring server for overwork
I0602 22:47:44.659548       7 monitoring.go:53] Starting monitoring server at :6601
I0602 22:47:44.659579       7 monitoring.go:75] Prometheus metric is enabled at :6601/worker/metrics
I0602 22:47:44.686105       7 storage.go:25] Warn: Failed to create client: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
I0602 22:47:44.686732       7 worker.go:154] Listening at port: localhost: 9000
I0602 22:47:44.687428       7 handlers.go:95] Worker connecting to coordinator: ws://localhost:8000/wso?zone=
I0602 22:47:44.692237       6 handlers.go:87] Coordinator: A worker is connecting...
I0602 22:47:44.693182       6 worker.go:41] Worker c7a8297a-cb1a-42a8-8850-80aad3490e46] Generated worker ID
I0602 22:47:44.693386       6 worker.go:41] Worker c7a8297a-cb1a-42a8-8850-80aad3490e46] Address:127.0.0.1
I0602 22:47:44.693555       6 worker.go:36] Worker c7a8297a-cb1a-42a8-8850-80aad3490e46] Is public: false zone: 
I0602 22:47:44.693715       7 handlers.go:72] Connected to coordinator successfully. &{0xc0003a61c0} <nil>
I0602 22:47:44.701944       7 coordinator.go:45] Received serverID  c7a8297a-cb1a-42a8-8850-80aad3490e46`

Docker Logs after launching localhost:8000 for the first time
`I0602 22:50:25.788816       7 main.go:37] Initializing coordinator server
I0602 22:50:25.788883       7 main.go:38] Coordinator configs {8800 http://localhost:8000   {6601 /coordinator false false}}
I0602 22:50:25.789079       7 coordinator.go:103] Initializing Coordinator Server
I0602 22:50:25.789075       7 coordinator.go:45] Starting monitoring server for coordinator
I0602 22:50:25.789140       7 monitoring.go:81] Monitoring server is disabled via config
I0602 22:50:25.792917       8 main.go:29] Initializing worker server
I0602 22:50:25.793407       8 worker.go:47] Starting monitoring server for overwork
I0602 22:50:25.793472       8 monitoring.go:53] Starting monitoring server at :6601
I0602 22:50:25.793484       8 monitoring.go:75] Prometheus metric is enabled at :6601/worker/metrics
I0602 22:50:25.797254       8 storage.go:25] Warn: Failed to create client: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
I0602 22:50:25.797369       8 worker.go:154] Listening at port: localhost: 9000
I0602 22:50:25.797508       8 handlers.go:95] Worker connecting to coordinator: ws://localhost:8000/wso?zone=
I0602 22:50:25.799349       7 handlers.go:87] Coordinator: A worker is connecting...
I0602 22:50:25.799489       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Generated worker ID
I0602 22:50:25.799514       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Address:127.0.0.1
I0602 22:50:25.799539       7 worker.go:36] Worker db35d0ea-1267-44cc-abad-1b12588185be] Is public: false zone: 
I0602 22:50:25.799818       8 handlers.go:72] Connected to coordinator successfully. &{0xc00012e340} <nil>
I0602 22:50:25.800488       8 coordinator.go:45] Received serverID  db35d0ea-1267-44cc-abad-1b12588185be
I0602 22:50:28.544908       7 handlers.go:161] Coordinator: A user is connecting...
I0602 22:50:28.545083       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Generated worker ID
I0602 22:50:28.545132       7 browser.go:30] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Get Room  Zone  From URL /ws?room_id=&zone=
I0602 22:50:28.545192       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Send sync[http://localhost:9000/echo]http://localhost:9000/echo
I0602 22:50:28.592340       8 worker.go:75] &{0xc000398320 0xc00032c400 {} 0x4e4370 false false false false 0xc000154600 {0xc0003b0000 map[] false false} map[Access-Control-Allow-Origin:[*]] true 0 -1 0 false false [] 0 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] [0 0 0 0 0 0 0 0 0 0] [0 0 0] 0xc000396070 0} echo
I0602 22:50:28.600730       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Latency mapmap[0xc0001bc120:12]
I0602 22:50:28.621099       7 browser.go:30] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received initwebrtc request -> relay to worker: db35d0ea-1267-44cc-abad-1b12588185be
I0602 22:50:28.621727       8 coordinator.go:57] Received a request to createOffer from browser via coordinator
I0602 22:50:28.622090       8 webrtc.go:115] === StartClient ===
I0602 22:50:28.679160       8 webrtc.go:135] Add video track
I0602 22:50:28.679869       8 webrtc.go:224] Created Offer
I0602 22:50:28.680467       8 coordinator.go:89] Start peerconnection 0b0709bd-8d76-4423-a1e0-f2b9308d1b95
I0602 22:50:28.680773       8 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 2130706431 172.17.0.2 42985 typ host
I0602 22:50:28.680979       8 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 1694498815 192.150.10.200 49239 typ srflx raddr 0.0.0.0 rport 54310
I0602 22:50:28.681189       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received SDP from worker -> sending back to browser
I0602 22:50:28.681308       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Received IceCandidate from worker -> relay to browser
I0602 22:50:28.681472       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Received IceCandidate from worker -> relay to browser
I0602 22:50:28.682105       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Received IceCandidate from worker -> relay to browser
I0602 22:50:28.700084       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received browser answered SDP -> relay to worker
I0602 22:50:28.701023       8 coordinator.go:106] Received answer SDP from browser
I0602 22:50:28.701475       8 webrtc.go:257] Set Remote Description
I0602 22:50:28.702007       8 webrtc.go:170] ICE Connection State has changed: checking
I0602 22:50:28.705253       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received IceCandidate from browser -> relay to worker
I0602 22:50:28.705857       8 coordinator.go:127] Received remote Ice Candidate from browser
I0602 22:50:28.705820       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received IceCandidate from browser -> relay to worker
I0602 22:50:28.705985       8 webrtc.go:268] Decoded Ice: candidate:3792981046 1 udp 2122260223 10.0.0.26 58969 typ host generation 0 ufrag 1dKY network-id 1 network-cost 10
I0602 22:50:28.706147       8 webrtc.go:276] Add Ice Candidate: candidate:3792981046 1 udp 2122260223 10.0.0.26 58969 typ host generation 0 ufrag 1dKY network-id 1 network-cost 10
I0602 22:50:28.706875       8 coordinator.go:127] Received remote Ice Candidate from browser
I0602 22:50:28.707222       8 webrtc.go:268] Decoded Ice: candidate:4169346470 1 udp 2122194687 10.177.44.112 51376 typ host generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.707433       8 webrtc.go:276] Add Ice Candidate: candidate:4169346470 1 udp 2122194687 10.177.44.112 51376 typ host generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.751873       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received IceCandidate from browser -> relay to worker
I0602 22:50:28.752399       8 coordinator.go:127] Received remote Ice Candidate from browser
I0602 22:50:28.752473       8 webrtc.go:268] Decoded Ice: candidate:2043357970 1 udp 1685987071 192.150.10.200 58717 typ srflx raddr 10.177.44.112 rport 51376 generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.752615       8 webrtc.go:276] Add Ice Candidate: candidate:2043357970 1 udp 1685987071 192.150.10.200 58717 typ srflx raddr 10.177.44.112 rport 51376 generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.753674       8 webrtc.go:170] ICE Connection State has changed: connected
I0602 22:50:28.753800       8 webrtc.go:174] ConnectionStateConnected
I0602 22:50:28.753841       8 webrtc.go:308] Start streaming
I0602 22:50:28.820028       8 webrtc.go:154] Data channel 'game-input'-'824635108826' open.
I0602 22:50:28.838337       8 webrtc.go:203] Received Voice from Client`

Launching from browser again given the following logs
`I0602 22:50:25.788816       7 main.go:37] Initializing coordinator server
I0602 22:50:25.788883       7 main.go:38] Coordinator configs {8800 http://localhost:8000   {6601 /coordinator false false}}
I0602 22:50:25.789079       7 coordinator.go:103] Initializing Coordinator Server
I0602 22:50:25.789075       7 coordinator.go:45] Starting monitoring server for coordinator
I0602 22:50:25.789140       7 monitoring.go:81] Monitoring server is disabled via config
I0602 22:50:25.792917       8 main.go:29] Initializing worker server
I0602 22:50:25.793407       8 worker.go:47] Starting monitoring server for overwork
I0602 22:50:25.793472       8 monitoring.go:53] Starting monitoring server at :6601
I0602 22:50:25.793484       8 monitoring.go:75] Prometheus metric is enabled at :6601/worker/metrics
I0602 22:50:25.797254       8 storage.go:25] Warn: Failed to create client: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
I0602 22:50:25.797369       8 worker.go:154] Listening at port: localhost: 9000
I0602 22:50:25.797508       8 handlers.go:95] Worker connecting to coordinator: ws://localhost:8000/wso?zone=
I0602 22:50:25.799349       7 handlers.go:87] Coordinator: A worker is connecting...
I0602 22:50:25.799489       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Generated worker ID
I0602 22:50:25.799514       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Address:127.0.0.1
I0602 22:50:25.799539       7 worker.go:36] Worker db35d0ea-1267-44cc-abad-1b12588185be] Is public: false zone: 
I0602 22:50:25.799818       8 handlers.go:72] Connected to coordinator successfully. &{0xc00012e340} <nil>
I0602 22:50:25.800488       8 coordinator.go:45] Received serverID  db35d0ea-1267-44cc-abad-1b12588185be
I0602 22:50:28.544908       7 handlers.go:161] Coordinator: A user is connecting...
I0602 22:50:28.545083       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Generated worker ID
I0602 22:50:28.545132       7 browser.go:30] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Get Room  Zone  From URL /ws?room_id=&zone=
I0602 22:50:28.545192       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Send sync[http://localhost:9000/echo]http://localhost:9000/echo
I0602 22:50:28.592340       8 worker.go:75] &{0xc000398320 0xc00032c400 {} 0x4e4370 false false false false 0xc000154600 {0xc0003b0000 map[] false false} map[Access-Control-Allow-Origin:[*]] true 0 -1 0 false false [] 0 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] [0 0 0 0 0 0 0 0 0 0] [0 0 0] 0xc000396070 0} echo
I0602 22:50:28.600730       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Latency mapmap[0xc0001bc120:12]
I0602 22:50:28.621099       7 browser.go:30] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received initwebrtc request -> relay to worker: db35d0ea-1267-44cc-abad-1b12588185be
I0602 22:50:28.621727       8 coordinator.go:57] Received a request to createOffer from browser via coordinator
I0602 22:50:28.622090       8 webrtc.go:115] === StartClient ===
I0602 22:50:28.679160       8 webrtc.go:135] Add video track
I0602 22:50:28.679869       8 webrtc.go:224] Created Offer
I0602 22:50:28.680467       8 coordinator.go:89] Start peerconnection 0b0709bd-8d76-4423-a1e0-f2b9308d1b95
I0602 22:50:28.680773       8 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 2130706431 172.17.0.2 42985 typ host
I0602 22:50:28.680979       8 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 1694498815 192.150.10.200 49239 typ srflx raddr 0.0.0.0 rport 54310
I0602 22:50:28.681189       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received SDP from worker -> sending back to browser
I0602 22:50:28.681308       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Received IceCandidate from worker -> relay to browser
I0602 22:50:28.681472       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Received IceCandidate from worker -> relay to browser
I0602 22:50:28.682105       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Received IceCandidate from worker -> relay to browser
I0602 22:50:28.700084       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received browser answered SDP -> relay to worker
I0602 22:50:28.701023       8 coordinator.go:106] Received answer SDP from browser
I0602 22:50:28.701475       8 webrtc.go:257] Set Remote Description
I0602 22:50:28.702007       8 webrtc.go:170] ICE Connection State has changed: checking
I0602 22:50:28.705253       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received IceCandidate from browser -> relay to worker
I0602 22:50:28.705857       8 coordinator.go:127] Received remote Ice Candidate from browser
I0602 22:50:28.705820       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received IceCandidate from browser -> relay to worker
I0602 22:50:28.705985       8 webrtc.go:268] Decoded Ice: candidate:3792981046 1 udp 2122260223 10.0.0.26 58969 typ host generation 0 ufrag 1dKY network-id 1 network-cost 10
I0602 22:50:28.706147       8 webrtc.go:276] Add Ice Candidate: candidate:3792981046 1 udp 2122260223 10.0.0.26 58969 typ host generation 0 ufrag 1dKY network-id 1 network-cost 10
I0602 22:50:28.706875       8 coordinator.go:127] Received remote Ice Candidate from browser
I0602 22:50:28.707222       8 webrtc.go:268] Decoded Ice: candidate:4169346470 1 udp 2122194687 10.177.44.112 51376 typ host generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.707433       8 webrtc.go:276] Add Ice Candidate: candidate:4169346470 1 udp 2122194687 10.177.44.112 51376 typ host generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.751873       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Received IceCandidate from browser -> relay to worker
I0602 22:50:28.752399       8 coordinator.go:127] Received remote Ice Candidate from browser
I0602 22:50:28.752473       8 webrtc.go:268] Decoded Ice: candidate:2043357970 1 udp 1685987071 192.150.10.200 58717 typ srflx raddr 10.177.44.112 rport 51376 generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.752615       8 webrtc.go:276] Add Ice Candidate: candidate:2043357970 1 udp 1685987071 192.150.10.200 58717 typ srflx raddr 10.177.44.112 rport 51376 generation 0 ufrag 1dKY network-id 2 network-cost 50
I0602 22:50:28.753674       8 webrtc.go:170] ICE Connection State has changed: connected
I0602 22:50:28.753800       8 webrtc.go:174] ConnectionStateConnected
I0602 22:50:28.753841       8 webrtc.go:308] Start streaming
I0602 22:50:28.820028       8 webrtc.go:154] Data channel 'game-input'-'824635108826' open.
I0602 22:50:28.838337       8 webrtc.go:203] Received Voice from Client
I0602 22:50:49.765989       7 cws.go:155] [!] read: read tcp 127.0.0.1:8000->127.0.0.1:40638: i/o timeout
I0602 22:50:49.766046       7 worker.go:41] Worker db35d0ea-1267-44cc-abad-1b12588185be] Unregister worker from coordinator
I0602 22:52:13.349315       7 cws.go:155] [!] read: websocket: close 1001 (going away)
I0602 22:52:13.349756       7 browser.go:35] Browser 0b0709bd-8d76-4423-a1e0-f2b9308d1b95] Disconnect from coordinator
I0602 22:52:13.417429       7 handlers.go:161] Coordinator: A user is connecting...
I0602 22:52:13.417856       7 browser.go:35] Browser 21f815c5-55bb-4b8e-ae1b-deb22a7251bb] Generated worker ID
I0602 22:52:13.417896       7 browser.go:30] Browser 21f815c5-55bb-4b8e-ae1b-deb22a7251bb] Get Room  Zone  From URL /ws?room_id=&zone=
I0602 22:52:13.417980       7 handlers.go:273] No server found`

**Comment by sergystepanov** [comment]
@akh64bit, thanks.
With these logs I can't really tell what's going on at which point, could you post one continuous log for the period of time starting from container run, then game run, then browser reload, then no games screen?

@giongto35, maybe you have any idea with those logs why socket has timeout?

**Comment by akh64bit** [comment]
Thanks @sergystepanov @giongto35 for having a look at the issue.

**Comment by sergystepanov** [comment]
> Oh Thank you @akh64bit so much. I indeed can reproduce with docker build. I will try to investigate it.
> @sergystepanov you can try it with `make dev.run-docker`. I spawned both worker and container in the same docker : D. And yes, this issue seems like the connection between worker and coordinator get crashed
> "I0602 22:50:49.765989 7 cws.go:155] [!] read: read tcp 127.0.0.1:8000->127.0.0.1:40638: i/o timeout"
> "I0602 22:52:13.349315 7 cws.go:155] [!] read: websocket: close 1001 (going away)"

Docker never worked for me:
```
I0604 10:44:36.475602       6 main.go:27] Initializing coordinator server
I0604 10:44:36.475735       6 main.go:28] Coordinator configs {8800 http://localhost:8000   {6601 /coordinator false false}}
I0604 10:44:36.475826       6 coordinator.go:45] Starting monitoring server for coordinator
I0604 10:44:36.475856       6 coordinator.go:103] Initializing Coordinator Server
I0604 10:44:36.475860       6 monitoring.go:81] Monitoring server is disabled via config
I0604 10:44:36.493570       7 main.go:29] Initializing worker server
I0604 10:44:36.496255       7 worker.go:47] Starting monitoring server for overwork
I0604 10:44:36.496322       7 monitoring.go:53] Starting monitoring server at :6601
I0604 10:44:36.496373       7 monitoring.go:75] Prometheus metric is enabled at :6601/worker/metrics
I0604 10:44:36.503159       7 storage.go:25] Warn: Failed to create client: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
I0604 10:44:36.503255       7 worker.go:154] Listening at port: localhost: 9000
I0604 10:44:36.503413       7 handlers.go:95] Worker connecting to coordinator: ws://localhost:8000/wso?zone=
I0604 10:44:36.504356       6 handlers.go:87] Coordinator: A worker is connecting...
I0604 10:44:36.504673       6 worker.go:41] Worker b15bf525-c599-4966-b7ef-2d5c58043400] Generated worker ID
I0604 10:44:36.504738       7 handlers.go:72] Connected to coordinator successfully. &{0xc00013e440} <nil>
I0604 10:44:36.504791       6 worker.go:41] Worker b15bf525-c599-4966-b7ef-2d5c58043400] Address:127.0.0.1
I0604 10:44:36.504861       6 worker.go:36] Worker b15bf525-c599-4966-b7ef-2d5c58043400] Is public: false zone: 
I0604 10:44:36.505344       7 coordinator.go:45] Received serverID  b15bf525-c599-4966-b7ef-2d5c58043400
I0604 10:44:51.744979       6 handlers.go:161] Coordinator: A user is connecting...
I0604 10:44:51.745119       6 browser.go:35] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Generated worker ID
I0604 10:44:51.745181       6 browser.go:30] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Get Room  Zone  From URL /ws?room_id=&zone=
I0604 10:44:51.745244       6 browser.go:35] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Send sync[http://localhost:9000/echo]http://localhost:9000/echo
I0604 10:44:51.815963       7 worker.go:75] &{0xc0003115e0 0xc0000fa500 {} 0x4e4370 false false false false 0xc000312f80 {0xc0003c4000 map[] false false} map[Access-Control-Allow-Origin:[*]] true 0 -1 0 false false [] 0 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] [0 0 0 0 0 0 0 0 0 0] [0 0 0] 0xc000086230 0} echo
I0604 10:44:51.846526       6 browser.go:35] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Latency mapmap[0xc000110720:37]
I0604 10:44:51.864765       6 browser.go:30] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Received initwebrtc request -> relay to worker: b15bf525-c599-4966-b7ef-2d5c58043400
I0604 10:44:51.865160       7 coordinator.go:57] Received a request to createOffer from browser via coordinator
I0604 10:44:51.865280       7 webrtc.go:115] === StartClient ===
I0604 10:44:51.929312       7 webrtc.go:135] Add video track
I0604 10:44:51.929746       7 webrtc.go:224] Created Offer
I0604 10:44:51.930005       7 coordinator.go:89] Start peerconnection 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42
I0604 10:44:51.930115       7 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 2130706431 172.17.0.2 39337 typ host
I0604 10:44:51.930252       7 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 1694498815 95.30.109.77 54059 typ srflx raddr 0.0.0.0 rport 47257
I0604 10:44:51.930485       6 worker.go:41] Worker b15bf525-c599-4966-b7ef-2d5c58043400] Received IceCandidate from worker -> relay to browser
I0604 10:44:51.930539       6 browser.go:35] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Received SDP from worker -> sending back to browser
I0604 10:44:51.930604       6 worker.go:41] Worker b15bf525-c599-4966-b7ef-2d5c58043400] Received IceCandidate from worker -> relay to browser
I0604 10:44:51.930635       6 worker.go:41] Worker b15bf525-c599-4966-b7ef-2d5c58043400] Received IceCandidate from worker -> relay to browser
I0604 10:44:52.012353       6 browser.go:35] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Received browser answered SDP -> relay to worker
I0604 10:44:52.013014       7 coordinator.go:106] Received answer SDP from browser
I0604 10:44:52.013305       7 webrtc.go:257] Set Remote Description
I0604 10:44:52.013446       7 webrtc.go:170] ICE Connection State has changed: checking
I0604 10:44:52.034205       6 browser.go:35] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Received IceCandidate from browser -> relay to worker
I0604 10:44:52.034562       7 coordinator.go:127] Received remote Ice Candidate from browser
I0604 10:44:52.034675       7 webrtc.go:268] Decoded Ice: candidate:1910078557 1 udp 2113937151 c54712b0-76ce-41f3-9ef5-5289a119212e.local 54062 typ host generation 0 ufrag mv6A network-cost 999
I0604 10:44:52.034860       7 webrtc.go:276] Add Ice Candidate: candidate:1910078557 1 udp 2113937151 c54712b0-76ce-41f3-9ef5-5289a119212e.local 54062 typ host generation 0 ufrag mv6A network-cost 999
I0604 10:44:52.058023       6 browser.go:35] Browser 8c2de6ff-ae73-4455-8f3b-76d56a7b9a42] Received IceCandidate from browser -> relay to worker
I0604 10:44:52.058374       7 coordinator.go:127] Received remote Ice Candidate from browser
I0604 10:44:52.058446       7 webrtc.go:268] Decoded Ice: candidate:842163049 1 udp 1677729535 95.30.109.77 54062 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag mv6A network-cost 999
I0604 10:44:52.058539       7 webrtc.go:276] Add Ice Candidate: candidate:842163049 1 udp 1677729535 95.30.109.77 54062 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag mv6A network-cost 999
I0604 10:45:02.015815       7 webrtc.go:170] ICE Connection State has changed: failed
```

**Comment by sergystepanov** [comment]
Also it's probably better to build proper canonical Docker images here, at gh https://github.com/giongto35/cloud-game/packages?package_type=Docker

**Comment by giongto35** [comment]
I just made an investigation. 
1) Try running the container few times subsequently, the app log stucks at different places.
2) Try spawning the program inside docker, cannot interrupt when the program got frozen => The container got malfunctioned.
3) Check CPU usage, it was 100%

=> It is due to high CPU usage of my new bad implementation. Container's CPU is capped, so everything inside is frozen.
I think your new Dockerfile should work now as well @sergystepanov 

Fixed here: https://github.com/giongto35/cloud-game/pull/191
